### ISSUE TITLE
tracking bug: add security headers

### ISSUE THREAD

**Issue by mrdomino** [urbit/urbit#4618]
**Problem**

A bunch of people spent a long time creating a really bad protocol for sending data back and forth over the internet. Part of this process involved creating a number of coping mechanisms for dealing with how bad the situation was. If Landscape makes use of this protocol, it should probably study and use those coping mechanisms where appropriate.

One of the coping mechanisms people added in at some point was the concept of “security headers”. Examples: `Content-Security-Policy`, `X-Frame-Options`, `X-XSS-Protection`, `X-Content-Type-Options`, `Referrer-Policy`, maybe even `Strict-Transport-Security`. If you use these, your browser may try somewhat harder not to do dumb things that open you up to security vulnerabilities.

**Proposed Solution**

Send appropriate security headers that both don't break Landscape functionality, and also don't allow those bad guys to break into your urbit.

**Possible Alternatives**

The main alternative is, keep being insecure and fix stuff gradually as people keep creating cleverer proof-of-concept exploits and posting them in Urbit Community.

**Additional context**

Possibly helpful resources:
* https://securityheaders.com/
* https://csp-evaluator.withgoogle.com/

E.g., securityheaders.com gives my Urbit dev instance an F:

<img width="476" alt="Screen Shot 2021-03-16 at 2 52 16 PM" src="https://user-images.githubusercontent.com/23019/111384850-368e9700-8667-11eb-9eb2-a0bcff8dafd9.png">

**Comment by Fang-** [comment]
tldr: eyre should maybe do a few of these things, but for the most part this is landscape/fileserver territory.

Eyre tries to be a dumb web server. Various bits of browser-specific behavior have crept in, but there's notable consistency across those: both cookie-based authentication and CORS handling are _easily and unambiguously detectable_ in incoming requests, and benefit greatly from _consistent, ship-wide_ state and behavior.  
You can almost imagine handling those things in userspace, except they're commonly desired in combination with Eyre-interal bindings (`%channel`, `%scry`).

Most of the headers suggested by these sites (Content-Security-Policy, X-Frame-Options, Referrer-Policy, Feature-Policy) are coupled very tightly to browser functionality, and don't really make sense to send with not-a-webpage responses. Eyre is not in the business of distinguishing between the ir/relevant cases, so the underlying handler should set the headers it deems appropriate. The only Eyre-side change here might be including some of these headers in the serving of the login page.

There's two headers that might be exceptions: Strict-Transport-Security and X-Content-Type-Options.

Strict-Transport-Security: when a request comes in over HTTPS, we could include this in the eventual response, to signal our commitment to the secure connection. If you're capable of setting up HTTPS, you're capable of keeping it up.  
Does this affect non-browser clients? Probably not. But at least it's easy to determine when to send this.

X-Content-Type-Options: its very existence is, of course, ridiculous. But saying "yes the given content-type is real" all of the time isn't strictly worse than praying the client is well-behaved.  
It's unclear to me what the security benefits of this would be, though.

**Comment by matildepark** [comment]
Considering the F grade was on the login page, these do at least need to be on that page.

The next tasks here are to
- deliberate what headers should be integrated and which ones where;
- stage it in a PR where we can test it against the Landscape build process;
- merge it into next-sys or next-userspace as appropriate

The problem is that I kind of don't know who has bandwidth for this. @ixv would you be willing to meet with @Fang- and draft something, then just implement some headers in file-server if necessary?

**Comment by brendanhay** [comment]
> press F to pay respects

**Comment by brendanhay** [comment]
![ship](https://user-images.githubusercontent.com/4493/111991837-0c305400-8b15-11eb-92ac-5dba0a699bb0.png)

**Comment by brendanhay** [comment]
So per the image above I modified all response through the reverse proxy to include the recommended headers for a hosted ship.

Two issues:

1. `script-src 'unsafe-inline'` has to be used because the proxy isn't (sanely) able to annotate the inline response tags like `<script nonce="nonce-<base64-value" />`. One way to avoid this is to calculate and assign randomised nonces per the eyre/landscape discussion above, or, the build/deploy pipeline could embed hashes computed at build time like `<script nonce="<hash-algorithm>-<base64-value>" />`.

2. After a cursory reading of the [Content-Security-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) reference I was unable to figure out how to get data-uri to behave, example: 

`Content Security Policy: The page’s settings blocked the loading of a resource at data:@file/ogg;base64,T2dnUwACAAAA… (“default-src”).`

**Comment by mrdomino** [comment]
Re: 2, `default-src data:` should work I believe.

**Comment by mrdomino** [comment]
revisiting this in a post-grid world: we now have all sorts of javascripts being included by third-party devs in unprincipled ways.

i have thus far been dealing with this using a whitelist of hashes and adding to it every time i install or update an app. this is obviously untenable.

if anyone wants to work on this, the following may be useful background: https://www.w3.org/TR/CSP3/#strict-dynamic-usage

tldr seems to be that the approved safety measure is to use nonces, and also include `'strict-dynamic'` for `script-src`. though apparently now they want you to get a booster every few months…

**Comment by zalberico** [comment]
@Fang- it's not obvious to me what the current state of this issue is. It looks like we have things set on hosting, but is there a specific action here otherwise? If so, I'd like to focus it on what those specific things are, if not I'd close.

**Comment by Fang-** [comment]
I think the tail part of my response here still stands as potential action-items for Eyre:

> Strict-Transport-Security: when a request comes in over HTTPS, we could include this in the eventual response, to signal our commitment to the secure connection. If you're capable of setting up HTTPS, you're capable of keeping it up.
Does this affect non-browser clients? Probably not. But at least it's easy to determine when to send this.
>
> X-Content-Type-Options: its very existence is, of course, ridiculous. But saying "yes the given content-type is real" all of the time isn't strictly worse than praying the client is well-behaved.
It's unclear to me what the security benefits of this would be, though.

**Comment by darighost** [comment]
> I think the tail part of my response here still stands as potential action-items for Eyre:
> ...

It would be good if we could add `data:` to the `default-src` directive of the Content Security Policy, which would make it so base64 images work, per https://github.com/urbit/urbit/issues/4618#issuecomment-804190243.

Unfortunately, I cannot even figure out where Eyre _defines_ its Content Security Policy, otherwise I'd submit a PR for this myself! Does anyone know?

**Comment by pkova** [comment]
Eyre does not define any CSP headers, but looks like some kind of reverse proxy in front of your ship on Tlon hosting does.

**Comment by patosullivan** [comment]
Yep, that's a tlon hosting thing.

**Comment by mrdomino** [comment]
I just set up security headers for a self-hosted ship, and I noticed that I had to separately add `data:` to `image-src` as well as `default-src`. So that's probably what's going on.
